[SPARK-10082] [MLlib] Validate i, j in apply DenseMatrices and SparseMatrices

MechCoder opened the pull request:
Given row_ind should be less than the number of rows
Given col_ind should be less than the number of cols.

The current code in master gives unpredictable behavior for such cases.

MechCoder committed:
[SPARK-10082] [MLlib] Validate i, j in apply (Dense and Sparse Matrices

MechCoder commented:
10k JIRA's wohoo :D

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41115 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41115/consoleFull) for   PR 8271 at commit [`af83250`](https://github.com/apache/spark/commit/af83250dee367cde5a1d8abaeea6fd0b444da0e8).

SparkQA commented:
[Test build #41115 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41115/console) for   PR 8271 at commit [`af83250`](https://github.com/apache/spark/commit/af83250dee367cde5a1d8abaeea6fd0b444da0e8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41115/
Test PASSed.

feynmanliang commented:
:+1: at first I was concerned that this might be unnecessary since `Array` checks for bounds, but this PR fixes a bug which occurs when a user provides an OOB row index and `values` causes overflow to the next column (assuming row-major order)

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala:
```diff
@@ -281,7 +281,11 @@ class DenseMatrix(
    */
-  override def apply(i: Int, j: Int): Double = values(index(i, j))
+  override def apply(i: Int, j: Int): Double = {
+    require(i < numRows && i >=0, s"Expected i to be non-negative and < $numRows, got $i")
```
nit: "Expected 0 <= i < $numRows bot got i=$i", same for others

feynmanliang replied:
Actually, perhaps we should move this check into `index` so other methods (e.g. `update`) will also get checked

feynmanliang commented on mllib/src/test/scala/org/apache/spark/mllib/linalg/MatricesSuite.scala:
```diff
@@ -74,6 +74,17 @@ class MatricesSuite extends SparkFunSuite {
     }
   }
 
+  test("apply in matrices incorrect input") {
```
After moving check to `index`, check `index` here instead of `apply`

MechCoder committed:
move check to index

MechCoder commented:
Actually this check is not necessary for SparseMatrices. But worst case it saves the effort of doing a binary search (and it should not return 0.0 if the row and col is not present)

I have addressed your comments/

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

MechCoder commented:
And I discovered the OOB bug by accident :)

SparkQA commented:
[Test build #41155 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41155/consoleFull) for   PR 8271 at commit [`cf7f509`](https://github.com/apache/spark/commit/cf7f5091174247ea945b8b4ae01eb02f64a07711).

SparkQA commented:
[Test build #41155 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41155/console) for   PR 8271 at commit [`cf7f509`](https://github.com/apache/spark/commit/cf7f5091174247ea945b8b4ae01eb02f64a07711).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41155/
Test PASSed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala:
```diff
@@ -281,9 +281,13 @@ class DenseMatrix(
    * @since 1.3.0
    */
-  override def apply(i: Int, j: Int): Double = values(index(i, j))
+  override def apply(i: Int, j: Int): Double = {
```
nit: this is a noop change, can we just keep the previous code?

feynmanliang commented:
minor suggestion but LGTM

CC @mengxr @jkbradley for final pass

MechCoder committed:
minor

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41159 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41159/consoleFull) for   PR 8271 at commit [`a92c382`](https://github.com/apache/spark/commit/a92c38287c273d82d3e22cf35ebc8216f33d0b2d).

SparkQA commented:
[Test build #41159 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41159/console) for   PR 8271 at commit [`a92c382`](https://github.com/apache/spark/commit/a92c38287c273d82d3e22cf35ebc8216f33d0b2d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41159/
Test PASSed.

mengxr commented:
I think this requires some micro-benchmark. I want to see the overhead of additional two checks. We can also test a single `require` statement that contains both checks.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala:
```diff
@@ -284,6 +284,8 @@ class DenseMatrix(
   override def apply(i: Int, j: Int): Double = values(index(i, j))
 
   private[mllib] def index(i: Int, j: Int): Int = {
+    require(i < numRows && i >=0, s"Expected 0 <= i < $numRows, got $i")
```
Nit: space after `>=`, and yes I think these can be combined into one check each. If there's an error I think it's helpful to log both i and j anyway.

mengxr commented:
Merged into master. I will send a follow-up PR to address minor issues.

MechCoder commented:
thanks !
